### PR TITLE
ASC-766 Add link generation facade

### PR DESCRIPTION
**Pull request by zreichert**
- adds a new facade to generate links to github test cases
- extends ZigZagTestLog class to use a LinkGenerationFacade object
- adds new properties to ZigZagTestLog

### CODE REVIEW

File: tests/unit/conftest.py [hunk 1]
```diff
@@ -12,24 +12,24 @@
...
+                <property name="PYTHON_ARTIFACT_SHA" value="Unknown"/>
+                <property name="APT_ARTIFACT_SHA" value="Unknown"/>
+                <property name="REPO_URL" value="https://github.com/rcbops/rpc-openstack"/>
+                <property name="JOB_NAME" value="PM_rpc-openstack-pike-rc-xenial_mnaio_no_artifacts-swift-system"/>
+                <property name="MOLECULE_TEST_REPO" value="molecule-validate-neutron-deploy"/>
+                <property name="MOLECULE_SCENARIO_NAME" value="default"/>
                 <property name="ci-environment" value="asc"/>
             </properties>
-    """
+    """  # noqa
```
Comment (johnduarte) [inline tests/unit/conftest.py]: The commit comment does not indicate why the test configuration property values have been changed from abstract names to specific values. Can you amend comment or split this into another commit to document the reason for this change?

File: tests/unit/test_cli.py [hunk 2]
```diff
@@ -18,7 +18,7 @@ def test_cli_happy_path(single_passing_xml, mocker):
     # Setup
     env_vars = {'QTEST_API_TOKEN': 'valid_token'}
     project_id = '12345'
-    test_cycle_name = 'RPC_PRODUCT_RELEASE'
+    test_cycle_name = 'pike'
```
Comment (johnduarte) [inline tests/unit/test_cli.py]: The commit comment does not indicate why this value has been changed. Can you indicate why?

File: tests/unit/test_link_generation_facade.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        zz.ci_environment = 'asc'
+        zz.testsuite_props = ASC_TESTSUITE_PROPS
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = FAILURE_OUTPUT
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == 'https://github.com/rcbops/molecule-validate-neutron-deploy/blob/pike-rc/molecule/default/tests/totally_real/verify_I_am_good_link.py#L23'  # noqa
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: The validation string here is too magical.

I do not know how the following values were derived:
* The github namespace expected for the `MOLECULE_TEST_REPO` defined in `ASC_TESTSUITE_PROPS`: `rcbops`.
* The molecule scenario name under test: `default`

The following value can be discovered, but is a bit obscure:
* The line number associated with the failure in the test: `23`.

File: tests/unit/test_link_generation_facade.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        zz.testsuite_props = ASC_TESTSUITE_PROPS
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = 'I dont contain an assert message'
+        zztl.def_line_number = '88'
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == 'https://github.com/rcbops/molecule-validate-neutron-deploy/blob/pike-rc/molecule/default/tests/totally_real/verify_I_am_good_link.py#L88'  # noqa
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: See comment for line 58.
The line number associated with the failure in this test is explicitly defined for the test_file mock. So, it is no longer mysterious.

File: tests/unit/test_link_generation_facade.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        zz.testsuite_props = ASC_TESTSUITE_PROPS
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = 'I dont contain an assert message'
+        zztl.def_line_number = ''
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == 'https://github.com/rcbops/molecule-validate-neutron-deploy/blob/pike-rc/molecule/default/tests/totally_real/verify_I_am_good_link.py'  # noqa
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: See comment for line 58.

FWIW, you could also break this line to fit within the flake8 line limit value using an expression to compose the URI.
```
assert failure_link == ('https://github.com/'
                        'rcbops/'
                        'molecule-validate-neutron-deploy/'
                        'blob/pike-rc/'
                        'molecule/default/'
                        'tests/totally_real/verify_I_am_good_link.py')
```

File: tests/unit/test_link_generation_facade.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        zz.ci_environment = 'mk8s'
+        zz.testsuite_props = MK8S_TESTSUITE_PROPS
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = FAILURE_OUTPUT
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == 'https://github.com/rcbops/mk8s/blob/master/tools/installer/tests/totally_real/verify_I_am_good_link.py#L23'  # noqa
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: See comment for line 58.

I do not know where the `tools/installer` path came from.

File: tests/unit/test_zigzag.py [hunk 7]
```diff
@@ -17,12 +17,12 @@
 # Globals
 # ======================================================================================================================
 # Shared expectations
-SHARED_TEST_LOG_EXP = {'build_url': 'BUILD_URL',
-                       'build_number': 'BUILD_NUMBER',
-                       'module_names': ['RPC_RELEASE',
-                                        'JOB_NAME',
-                                        'MOLECULE_TEST_REPO',
-                                        'MOLECULE_SCENARIO_NAME',
+SHARED_TEST_LOG_EXP = {'build_url': 'https://rpc.jenkins.cit.rackspace.net/job/PM_rpc-openstack-pike-rc-xenial_mnaio_no_artifacts-swift-system/78/',  # noqa
```
Comment (johnduarte) [inline tests/unit/test_zigzag.py]: Can you comment on the reason for changing these values?

File: tests/unit/test_link_generation_facade.py (hunk 6 above)
Comment (zreichert) [reply tests/unit/test_link_generation_facade.py]: I'll comment in the code for this one.  
Its specific to the mk8s repo and can not be derived from the XML at this time.

File: tests/unit/test_link_generation_facade.py (hunk 5 above)
Comment (zreichert) [reply tests/unit/test_link_generation_facade.py]: Fixed

File: tests/unit/test_zigzag.py (hunk 7 above)
Comment (zreichert) [reply tests/unit/test_zigzag.py]: These values where updated in the tests because just having a string with any value is not acceptable to use with this feature.  The feature depends on the values that exist inside the strings.  

@johnduarte Would a comment in the commit be acceptable?

Comment (zreichert) [reply tests/unit/test_zigzag.py]: I just pushed an update, please let me know if I have not adequately described whats going on.

Comment (johnduarte) [reply tests/unit/test_zigzag.py]: 👍 works for me.

File: tests/unit/conftest.py (hunk 1 above)
Comment (johnduarte) [reply tests/unit/conftest.py]: addressed in new commit message.

File: tests/unit/test_cli.py (hunk 2 above)
Comment (johnduarte) [reply tests/unit/test_cli.py]: addressed in new commit message.

File: tests/unit/test_link_generation_facade.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        zz.testsuite_props = ASC_TESTSUITE_PROPS
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = FAILURE_OUTPUT
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == ('https://github.com/'
+                                'rcbops/'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+                                'molecule-validate-neutron-deploy/'
+                                'blob/pike-rc/'
+                                'molecule/default'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = 'I dont contain an assert message'
+        zztl.def_line_number = '88'
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == ('https://github.com/'
+                                'rcbops/'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+                                'molecule-validate-neutron-deploy/'
+                                'blob/pike-rc/'
+                                'molecule/default'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = 'I dont contain an assert message'
+        zztl.def_line_number = ''
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == ('https://github.com/'
+                                'rcbops/'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+                                'molecule-validate-neutron-deploy/'
+                                'blob/pike-rc/'
+                                'molecule/default'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+        zz.testsuite_props = MK8S_TESTSUITE_PROPS
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = FAILURE_OUTPUT
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == ('https://github.com/'
+                                'rcbops/'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+                                'mk8s/'
+                                'blob/master/'
+                                'tools/installer/'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+        zz.testsuite_props = MK8S_PR_TESTSUITE_PROPS
+
+        zztl = mocker.MagicMock()
+        zztl.test_file = TEST_FILE
+        zztl.failure_output = FAILURE_OUTPUT
+
+        lgf = LinkGenerationFacade(zz)
+        failure_link = lgf.github_testlog_failure_link(zztl)
+        assert failure_link == ('https://github.com/'
+                                'zreichert/'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+                                'mk8s/'
+                                'blob/asc-123/master/this_is_my_feature/'
+                                'tools/installer/'
```
Comment (johnduarte) [inline tests/unit/test_link_generation_facade.py]: Don't know where this expected value is coming from.

File: tests/unit/test_link_generation_facade.py (hunk 8 above)
Comment (johnduarte) [reply tests/unit/test_link_generation_facade.py]: Defined in `REPO_URL` fed to `zz.testsuite_props` above.

File: tests/unit/test_link_generation_facade.py (hunk 10 above)
Comment (johnduarte) [reply tests/unit/test_link_generation_facade.py]: Defined in `REPO_URL` fed to `zz.testsuite_props` above.

File: tests/unit/test_link_generation_facade.py (hunk 12 above)
Comment (johnduarte) [reply tests/unit/test_link_generation_facade.py]: Defined in `REPO_URL` fed to `zz.testsuite_props` above.

File: tests/unit/test_link_generation_facade.py (hunk 14 above)
Comment (johnduarte) [reply tests/unit/test_link_generation_facade.py]: Defined in `REPO_URL` fed to `zz.testsuite_props` above.
